torch/testing/_internal/common_methods_invocations.py

```diff
@@ -3553,14 +3553,32 @@ def sample_inputs_expand_as(op_info, device, dtype, requires_grad, **kwargs):
              )
 
     def generator():
-        for case in cases:
-            shape, shape_other = case
+        for shape, shape_other in cases:
             yield(SampleInput(make_arg(shape, requires_grad=requires_grad),
                               args=(make_arg(shape_other, requires_grad=False), )))
 
     return list(generator())
 
 
+def sample_inputs_where(op_info, device, dtype, requires_grad, **kwargs):
+    make_arg = partial(make_tensor, dtype=dtype, device=device, requires_grad=requires_grad)
+    make_bool_mask = partial(make_tensor, dtype=torch.bool, device=device, requires_grad=requires_grad)
```

Comment:
The function this is replacing guaranteed that at least one value in the mask is True

This PR should preserve that behavior. It can do so by creating its own make_bool_mask function that checks if the tensor contains at least one True value and, if not, inserts a True a "True" at a random point, for example.

Reply:
Is it okay to use `mask_not_all_zero`?

Reply:
Sure; although eventually it'd be good to improve upon it and not just redraw

Reply:
@mruberry take a look at the updated comment. I am confused.

Reply:
I think some of the indexing ops do allow the indices (and maybe masks) to be on the CPU:

```
import torch

t = torch.tensor((1, 2, 3), device='cuda')
indices = torch.tensor((1,))
values = torch.tensor((4,), device='cuda')
t.index_put_((indices,), values)
: tensor([1, 4, 3], device='cuda:0')
```

Reply:
In the code it looks like the function first manipulates the tensors according to the indices argument, then calls TensorIterator on the input and values tensors only.

Reply:
Is this behaviour intended?

Reply:
I think so; and it would be BC-breaking to change

Reply:
How does this look?
```python
    def make_bool_mask(shape):
        # Make sure atleast one element is nonzero,
        # except for empty tensor
        mask_t = make_tensor(shape, dtype=torch.bool, device=device, requires_grad=False)
        if mask_t.numel() == 1 and len(mask_t.shape) > 0:
            mask_t[0] = True
            return mask_t
        elif reduce(lambda x, y: x*y, mask_t.shape, 1) == 0:
            return mask_t

        def random_index(shape):
            return tuple(map(lambda max_idx: random.randint(0, max_idx), shape))

        mask_t[random_index(mask_t.shape)] = True
        return mask_t
```

Reply:
The handling of tensors with zero elements looks good but I think that check can be simplified to `mask_t` having numel == 0? The handling of the general case also looks correct.

I think the tricky case, then, is if shape is `()` or `(1,)`. For these I think you need two checks, one for numel() == 1 and a then a second check for len(shape) == 0. In both these cases the initial construction of the tensor is irrelevant, because you know the tensor you want to return, so I would put these checks before the make_tensor case, short-circuit on 0 numel, and then have the random_index case guarded behind the call to sum()

Reply:
```
if numel == 1:
  if len(shape) == 0:
    return torch.tensor(True, ...)
  return torch.Tensor((True,), ...)

mask = make_tensor(...)
if mask.numel() == 0:
  return mask

if mask.sum() == 0:
  # sets a random index to True

return mask
```

Reply:
Tried to incorporate the suggestion
How about?
```python

    def make_bool_mask(shape):
        # Make sure atleast one element is nonzero,
        # except for empty tensor
        mask_t = make_tensor(shape, dtype=torch.bool, device=device, requires_grad=False)

        if mask_t.numel() == 0:
            return mask_t
        elif mask_t.numel() == 1:
            # Added these after make tensor so that
            # we have to specify `dtype`, `device`, etc only
            # once to make_tensor
            mask_t.fill_(True)
            return mask_t

        if mask_t.sum() == 0:
            def random_index(shape):
                return tuple(map(lambda max_idx: random.randint(0, max_idx), shape))

            mask_t[random_index(mask_t.shape)] = True
            return mask_t

        return mask_t

```

Reply:
LGTM!